Fixes #24041 - Move HW Models Table to pf-react

boaz0 opened the pull request:
This is almost ready - 

* Need to add tests to `Rails`. I will be working on it. In addition, I will add integration tests to `ModelsTable`.
* I did the best I can to make the table look similar to what it looked before. However, Katello implementation didn't fit here. Probably because I am not using `Pagination` and `Empty State` is rendered by Rails now.

* I didn't enter the pagination here, I will do that in the next PR. I wanted this PR to be smaller and easy to review as much as possible.

Thanks.

boaz0 committed:
Fixes #24041 - Move HW Models Table to pf-react

timogoebel commented on app/models/model.rb:
```diff
@@ -18,4 +18,22 @@ class Model < ApplicationRecord
   scoped_search :on => :hardware_model, :complete_value => :true
+
+  def hosts_count
+    Host::Managed.authorized.group(:model_id).count.fetch(id, 0)
```
Will this create a lot of n+1 queries?

boaz0 replied:
I will look into it. You might be right.

boaz0 replied:
I believe this means that there is no `N+1`:

```ruby
irb(main):119:0> Host::Managed.authorized.group(:model_id).count
D, [2018-07-02T16:57:23.009902 #4328] DEBUG -- :    (0.2ms)  SELECT COUNT(*) AS count_all, "hosts"."model_id" AS hosts_model_id FROM "hosts" WHERE "hosts"."type" IN ('Host::Managed') GROUP BY "hosts"."model_id"
=> {1=>11, 2=>1}
```

waldenraines commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,51 @@
+      columns, filterFnc, results, store,
+    } = this.props;
+    return (
+      <PfTable.PfProvider
```
There is also a foreman wrapper of the patternfly table that provides some additional functionality (such as integrating with the `EmptyState` component) in @amirfefer's about page PR (https://github.com/theforeman/foreman/pull/5024).

boaz0 replied:
Yes, I saw that PR. Definitely, will be useful when I turn `Hardware Models` into a `React` page. Although, we need to figure out best practices with `reactabular` since there are many ways to setup a table and each one of us did it differently due to different requirements/needs.

ohadlevy replied:
I would hope we have a common set of defaults that is mostly plug&pray (similar to how table css classes are defined today)

waldenraines commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableFormats.js:
```diff
@@ -0,0 +1,72 @@
+  (value.includes('delete') ? (
+    <span className="btn btn-sm btn-default">
+      <a
+        data-confirm={`Delete ${name}`}
```
This should be translated via `__()`.

waldenraines commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableSchema.js:
```diff
@@ -0,0 +1,100 @@
+}) => [
+  {
+    header: {
+      label: 'Name',
```
We need to translate all of these header labels via `__()`.

waldenraines reviewed (commented):
Code looks good overall to me, just a few comments.   Would be good to get @sharvit, @amirfefer, and/or @tstrachota to have a look as well.

ohadlevy commented on app/controllers/models_controller.rb:
```diff
@@ -6,6 +6,13 @@ class ModelsController < ApplicationController
+      @store[m.id] = {
+        :hosts_count => m.hosts_count,
+        :actions => m.actions_by_user(User.current)
+      }
```
this can be written as
```ruby
    @store = {}
    resource_base_search_and_page.each do |m|
      @store[m.id] = {
        :hosts_count => m.hosts_count,
        :actions => m.actions_by_user(User.current)
      }
    end
```

boaz0 replied:
`resource_base_search_and_page` gives only the first 20 results. I want `store` to contain all models' `hosts_counts` and `actions` and to display (in the future with `redux-reselect`) only the hosts given in the API `response` as you can see [here](https://github.com/ripcurld/foreman/blob/b4d8d29b5f8136cd178cb49def5ae5592d0ecaa9/webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js#L37)

ohadlevy commented on app/controllers/models_controller.rb:
```diff
@@ -6,6 +6,13 @@ class ModelsController < ApplicationController
+    @store = {}
+    Model.all().map do |m|
+      @store[m.id] = {
+        :hosts_count => m.hosts_count,
```
you can reuse the same methods that generate the statistics charts, e.g.
```ruby
Host::Managed.count_distribution(:model)
2018-07-01T10:48:19 [D|sql|]    (11.8ms)  SELECT COUNT(*) AS count_all, hosts.model_id AS hosts_model_id FROM "hosts" WHERE "hosts"."type" IN ('Host::Managed') GROUP BY hosts.model_id
2018-07-01T10:48:19 [D|sql|]   Model Load (2.4ms)  SELECT "models".* FROM "models" WHERE ("models"."id" IN (1, 2, 5, 10, 15, 17) OR "models"."id" IS NULL) ORDER BY models.name
[{:label=>"OptiPlex 780", :data=>1}, {:label=>"KVM", :data=>9}, {:label=>"3680B56", :data=>1}, {:label=>"IBM System x -[7870K4G]-", :data=>2}, {:label=>"RHEV Hypervisor", :data=>1}, {:label=>"VMware Virtual Platform", :data=>1}]
```

ohadlevy replied:
or as an "api" query at /statistics/manufacturer.json
result is:
```json
{
  "id": "manufacturer",
  "data": [
    ["Dell Inc.", 2],
    ["IBM", 2],
    ["LENOVO", 1],
    ["Red Hat", 12],
    ["Seabios", 1],
    ["VMware, Inc.", 1]
  ]
}
```

EDIT: just realized this is based on facter data, not foreman's model, leaving it here for reference but probably not a great way forward for this usage case.

ohadlevy replied:
or you can just run the count/group_by query yourself, it will save you an extra query or two

ohadlevy commented on app/models/model.rb:
```diff
@@ -18,4 +18,24 @@ class Model < ApplicationRecord
+      unless user.allowed_to?(
+        {
+          :controller => :models,
+          :action => action,
```
what happens once we delete the UI controller (or methods) once we use the API?

boaz0 replied:
I am not sure I understand what is the `UI` controller. Can you elaborate more?
Thanks :smiley:

ohadlevy replied:
e.g if we no longer need the models_controller and just use the official api?

ohadlevy replied:
my other concern here, is that this will quickly grow into a huge table in memory, if you keep allowed operations per object, I wonder, @waldenraines how does katello handle permissions today (e.g. to know if you can display a link or not for objects that have filters?)

boaz0 replied:
Yep, that's a good point. 

I thought maybe we would add this information into `api/models` or create another API end point to return models' data along with actions and hosts counts. Another option is to use `GQL`.

Anyway, I will look more into this and mention that here](https://community.theforeman.org/t/move-hardware-model-page-to-react/10070/24).

ohadlevy reviewed (commented):
basic comments on the rails side, didnt look at the js yet.

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,51 @@
+        striped
+        bordered
+        hover
+        columns={columns}
```
would be nice to see all of these become a default of some sort.

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,51 @@
+          body: {
+            cell: props => (
+              <PfTable.Cell className={props.className} {...props.data}>
+                {props.children}
```
missing table spinner while loading api results?

amirfefer replied:
You can use the LoadingState from [katello](https://github.com/Katello/katello/blob/master/webpack/move_to_pf/LoadingState/LoadingState.js)

ohadlevy replied:
there is also foreman's [Loader](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/common/Loader/index.js) - we should really consolidate all of these kind of components as its just a ui gaps (and bugs) waiting to happen between foreman and katello again....

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableFormats.js:
```diff
@@ -0,0 +1,72 @@
+import React from 'react';
+
+export const hostsCountFormat = (value, { rowData: { name } }) => (
+  <a href={`hosts?search=model+%3D+"${encodeURI(name)}"`}>{value}</a>
```
does it make sense to add as a generic component? I'm sure we will have other host counts in other places too

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableFormats.js:
```diff
@@ -0,0 +1,72 @@
+
+export const actionsFormat = (value, { rowData: { name, id } }) =>
+  (value.includes('delete') ? (
+    <span className="btn btn-sm btn-default">
```
:+1: to handle deleting via modal later on, but maybe we should use patternfly button component?

boaz0 replied:
Yes, this part is going to be done soon. `Katello` has `ConfirmDialog` that can be used here. 
I just didn't want to overload the number of changed files.

ohadlevy replied:
thats fine, we should probably replace all delete buttons across vs just change it here.

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableFormats.js:
```diff
@@ -0,0 +1,72 @@
+  ) : null);
+
+export const nameFormat = (value, { rowData: { actions, id, name } }) =>
+  (actions.includes('edit') ? (
```
does it make sense to have only one function that does a switch statement on actions, and call the relevant function? I would imagine in other pages we will have a lot more actions.

Also... these are probably generic actions? in that case maybe we should put them under the common namespace?

boaz0 replied:
Sounds good, I can work on it in this PR or in a future PR.

ohadlevy replied:
if its easy, i would move it now, if not, future PR is ok.

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableFormats.js:
```diff
@@ -0,0 +1,72 @@
+
+const orderAPI = {
+  ASC: 'DESC',
+  DESC: 'ASC',
```
?? our API is opposite ?

boaz0 replied:
The name is misleading. Basically, if now the order is "ASC" then in the next call we should set the order to "DESC" that's what I meant.

Anyway, I understand the confusion and will fix it.

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableReducer.js:
```diff
@@ -0,0 +1,42 @@
+      return state.set('status', STATUS.PENDING);
+    case MODELS_TABLE_SUCCESS:
+      return state
+        .set('status', STATUS.RESOLVED)
```
TBH  I'm not sure if it really matters for performance, but maybe consider merge(https://github.com/rtfeldman/seamless-immutable) instead of multiple sets?

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableSchema.js:
```diff
@@ -0,0 +1,100 @@
+    },
+    cell: {
+      props: {
+        className: 'ellipsis',
```
afair we have an ellipsis component? @amirfefer ?

boaz0 replied:
That's true, but looking at the DOM, it is different: now `<td>` has `ellipsis` in its class while using the  component a `<div>` is wrapped inside the `<td>` which wraps the value. If that is not a problem, I can use the component with no problems. :+1:

amirfefer replied:
> If that is not a problem

why it  could be a problem?

ohadlevy commented on app/views/models/index.html.erb:
```diff
@@ -2,32 +2,6 @@
-  </tbody>
-</table>
+<div id="models_table"></div>
+<%= mount_react_component("ModelsTable", "#models_table", {:url => "api/#{request.fullpath}", :store => @store}.to_json) %>
```
I find the term store a bit over used, maybe we can be more specific ? (or just flatten the store out?)

boaz0 replied:
No problem :+1:

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableFormats.js:
```diff
@@ -0,0 +1,72 @@
+  ));
+
+const orderChars = {
+  ASC: '▲',
```
looking at PF documentation at https://www.patternfly.org/pattern-library/

they have the following html for table headers:
```html
<th
  class="sorting_asc"
  tabindex="0"
  aria-controls="patternTable"
  rowspan="1"
  colspan="1"
  aria-sort="ascending"
  aria-label="Pattern: activate to sort column descending"
>
  Pattern
</th>
```

amirfefer commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,51 @@
+}
+
+ModelsTable.propTypes = {
+  columns: PropTypes.array.isRequired,
```
please use arrayOf for arrays, and shape for objects.
[This](https://www.fullstackreact.com/p/appendix-a-proptypes/#prop_type_array_of_type) could help

amirfefer commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableFormats.test.js:
```diff
@@ -0,0 +1,62 @@
+  });
+
+  it('formats a delete action button if authorized', () => {
+    expect(actionsFormat(['delete'], { rowData: { name: 'model13', id: 13 } })).toMatchSnapshot();
```
could you use prettifier?

amirfefer commented on webpack/assets/javascripts/react_app/components/ModelsTable/index.js:
```diff
@@ -0,0 +1,35 @@
+      sortBy,
+      sortOrder,
+    }),
+    filterFnc: (store = {}, displayModels = []) =>
```
afaik on each react's `render()` this function's ref will be changed, which might lead to poor performance. 
is there a reason to have this `filterFnc` here?

amirfefer reviewed (changes requested):
@ripcurld  Great progress, I left some inline comments.

Looks like there is no tooltip:
![screenshot from 2018-07-04 15-22-32](https://user-images.githubusercontent.com/11807069/42277293-8dadbf56-7f9f-11e8-8ce5-d93b9a34aef5.png)
you can use `EllipsisWithTooltip`

Foreman test failures are related: Please fix the hw-model integration test.

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableActions.js:
```diff
@@ -0,0 +1,43 @@
+export const updateSearch = (current) => {
+  const init = {
+    page: current.params && current.params.page,
+    perPage: current.params && current.params.per_page,
```
can you use the syntax of
```js
    const { params: { page, per_page, order, search }, table } = current;
```

ohadlevy replied:
or if you don't like that syntax, you can also:
```js
export const updateSearch = ({ params, table }) => {
  const init = {
    page: params && params.page,
    perPage: params && params.per_page,
    order: params && params.order,
    searchQuery: params && params.search,
    results: table,
  };
```

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,63 @@
+    ) : (
+      <Table
+        columns={schema({
+          url: new URI('models.json')
```
maybe this should be extracted into a function?

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,63 @@
+        })}
+        rows={results}
+        search={!!searchQuery}
+        emptyState={{
```
I would extract this too, just to make it easier to read

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,63 @@
+          documentation: {},
+          action: {
+            url: '#',
+            title: '',
```
? empty action?

boaz0 replied:
It will throw an error in the console if `action` is empty :disappointed:

ohadlevy commented on app/controllers/models_controller.rb:
```diff
@@ -6,6 +6,36 @@ class ModelsController < ApplicationController
+    order = (sort_param.length == 2) ? sort_param[1] : nil
+    @json_models = {
+      :results => @table,
+      :per_page => params.fetch(:per_page, "20"),
```
the defaults should not be defined here, but rather reuse settings.

boaz0 replied:
I agree with you. I am going to look how to wrap the `@table` response in a way it's done in the `api/v2` so the defaults will be taken from the `api/v2`.

ohadlevy commented on app/controllers/models_controller.rb:
```diff
@@ -6,6 +6,36 @@ class ModelsController < ApplicationController
+      :results => @table,
+      :per_page => params.fetch(:per_page, "20"),
+      :page => params.fetch(:page, "1"),
+      :total => @table.length.to_s,
```
why not use the API? it seems to have the exact set of data?

boaz0 replied:
`hosts_count` and `actions` aren't included in the `API` results and this information is needed when for example we have 6 models, 5 models are shown in each page and we are in page 1. When loading the page, only the first 5 models will have this information, when clicking on sorting by name in descanding, the API will return the last 5 models but the fist model's `actions` and `hosts_count` are missing which ends up with showing the first item as disabled and can't be deleted by the user even though it can be.

I don't know how I can wrap `@table` into `api/v2` response. I know it's available in `Api2::BaseController` or something like that but I don't know if that's possible. 
Anyway, I put it here to see if that format is good enough and improve this more.

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableFormats.js:
```diff
@@ -0,0 +1,60 @@
+        data-id={`aid_models_${id}-${encodeURI(name)}`}
+        rel="nofollow"
+        data-method="delete"
+        href={`/models/${id}-${encodeURI(name)}`}>
```
nitpick: please ensure you are using prettier, on my env it format this (and others) differently.

boaz0 replied:
Can you show me how it looks on your env after running `prettier`. Thanks.

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableFormats.test.js:
```diff
@@ -0,0 +1,62 @@
+    { url: 'api/models?page=2', order: 'ASC', name: 'name' },
+    { url: 'api/models?page=1', order: 'DESC', name: 'hardware_model' },
+  ].forEach((params) => {
+    expect(sortUrl(params.url, params.name, params.order)).toMatchSnapshot();
```
```js
].forEach(({ url, name, order }) => {
   expect(sortUrl(url, name, order)).toMatchSnapshot();
```

ohadlevy replied:
i good library to detect this kind of quirks is lebab.

timogoebel commented on app/controllers/models_controller.rb:
```diff
@@ -6,6 +6,36 @@ class ModelsController < ApplicationController
   def index
     @models = resource_base_search_and_page
+    @hosts_counts = Model.hosts_counts
+    @table = @models.map do |m|
```
This looks very hacky, why don't we retrieve the data via a separate API call? That would also support pagination etc.?

boaz0 replied:
I didn't do so because I was afraid that it might be a problem in regard to API versions. However, if it's OK, I am going to do it.

timogoebel commented on app/views/models/index.html.erb:
```diff
@@ -2,32 +2,8 @@
-</table>
+<div id="models_table"></div>
+<%= mount_react_component("ModelsTable", "#models_table", {
+    :params => request.query_parameters(),
```
Nit: You can omit the `()` here.

amirfefer replied:
I'd like to use  `react router` on HW model page, instead of using this server side mounting.
Though i'm totally fine if it wouldn't be in this PR, just food for thought.

ohadlevy commented:
I'm not sure I understand what "it" is but I would just suggest to fetch
the table data via the api while getting the permissions and host count
when mounting the component?

On Thu, Jul 5, 2018, 6:13 PM ripcurld <notifications@github.com> wrote:

> *@ripcurld* commented on this pull request.
> ------------------------------
>
> In app/controllers/models_controller.rb
> <https://github.com/theforeman/foreman/pull/5755#discussion_r200390253>:
>
> > @@ -6,6 +6,36 @@ class ModelsController < ApplicationController
>
>    def index
>      @models = resource_base_search_and_page
> +    @hosts_counts = Model.hosts_counts
> +    @table = @models.map do |m|
>
> I didn't do so because I was afraid that might be a problem in regard to
> API versions. However, if it's no a problem, I am going to do it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5755#discussion_r200390253>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx2_N7MufH0RQAxw3320lXPwMpl8xks5uDi0ugaJpZM4U8um9>
> .
>

boaz0 commented:
@ohadlevy - getting the permissions and hosts count when mounting the component could be a problem:

Let's say there are 26 models -> on mounting the component only the first 25 models' permissions and hosts count are passed to the component (by default 20 models are displayed in one page) -> user clicks on the "Name" header to sort models in descending -> api will give the last 25 models -> the component doesn't have permissions and hosts count information for one of the 25 models.

If permissions and hosts count are given from `api/models` or from `/models.json` then this solves my problem.

I didn't touch the `Api::V2` model controller because I was afraid that adding or changing it will require updating the API version to `v3` or something like that.

ohadlevy commented:
@ripcurld how about we add a json response for permissions, where we could ask the permissions of all of the objects we have? (either search term or list of id's or something like that?) (or maybe we can use the permission api as it is today?) /cc @timogoebel if you have another idea?

timogoebel commented:
> /cc @timogoebel if you have another idea?

I believe we should add more fields to the index api endpoint, something like `canEdit: true` or `canDelete: true`; or even `currentUserActions: ['delete', 'edit', 'show']`. That could be easily done both using graphql or our rest. I believe even quite generic. `if model includes authorizable, list allowed actions.`

Something like this: http://blog.rstankov.com/handling-user-permissions-in-graphql/

amirfefer commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.css:
```diff
@@ -0,0 +1,4 @@
+.tableloader {
```
Does it a general styling for a table, or specific to this HW model table?

boaz0 replied:
Probably to all tables. Although, I think `placeholders` like you see when you're loading `Youtube` is better in this case. Thanks for noticing this.

amirfefer commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableActions.js:
```diff
@@ -0,0 +1,43 @@
+export const sortTable = (url, controller) => dispatch =>
+  ajaxRequestAction({
+    dispatch,
+    requestAction: TABLE_REQUEST,
```
If this action is responsible for sorting, wouldn't it be better to use more informative consts?
something like `SORT_HW_MODEL_TABLE_REQUEST` ?

boaz0 replied:
Super right! I actually found that `patternfly-react` as such a constant and I am going to use it here.

amirfefer commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableActions.js:
```diff
@@ -0,0 +1,43 @@
+  };
+
+  const payload = Object.keys(init)
+    .filter(k => ![undefined, null].includes(init[k]) && k !== 'order')
```
afaik this filtering  shouldn't be in the action, I think it would be better to keep the redux state as is , and then use a selector via `mapsStateToProps` for filtering.

boaz0 replied:
SGTM, I am just waiting on [your PR](https://github.com/theforeman/foreman/pull/5778) before doing so.

amirfefer commented on webpack/assets/javascripts/react_app/components/ModelsTable/index.js:
```diff
@@ -0,0 +1,17 @@
+import reducer from './ModelsTableReducer';
+import * as actions from './ModelsTableActions';
+
+const mapStateToProps = ({ models }) => ({ ...models, schema });
```
Why keeping the schema in the props? if I understand it right, it doesn't part of the state.

boaz0 replied:
Again - you're correct. I don't think `schema` should be passed as props.
The `schema` has all the information on how the `Models Table` should be structured and thus I think it should be part of the component itself. I hope you agree with me too.

boaz0 commented:
@ohadlevy @timogoebel thanks a lot for your help.
I changed the `rabl` files to include `hosts_count` and `can*` attributes. 
To elaborate more, I added `app/views/api/v2/authorizable.json.rabl` that checks if the object has `Authorizable` mixing and if so checks if the current user can `edit`, `delete` and `create` the object's resource type.

Please let me know if I could make it better.

timogoebel commented on app/views/api/v2/authorizable.json.rabl:
```diff
@@ -0,0 +1,7 @@
+object @resource
+
+if @object.try(:class).try(:ancestors).to_a.include?(Authorizable)
```
I'm wondering if this should be guarded by a special parameter, something like [thin](https://github.com/theforeman/foreman/blob/d5f2f1de95e4cc665705c5047ccc89660f0bb92d/app/controllers/api/v2/hosts_controller.rb#L35-L52). Should we introduce this in a separate PR? That might make reviewing easier. Does this have tests?

timogoebel commented on app/views/api/v2/authorizable.json.rabl:
```diff
@@ -0,0 +1,7 @@
+object @resource
+
+if @object.try(:class).try(:ancestors).to_a.include?(Authorizable)
+    node(:canEdit) { |resource| resource.authorized?(:"#{:edit}_#{resource.class.name.pluralize.downcase}") } 
```
`#{:edit}` -> `edit` ;-)

But luckily we have a helper to get the permission name, `resource.permission_name('create')`.

Also please use snake case in the APIv2 for consistency.

timogoebel commented on app/views/api/v2/authorizable.json.rabl:
```diff
@@ -0,0 +1,7 @@
+
+if @object.try(:class).try(:ancestors).to_a.include?(Authorizable)
+    node(:canEdit) { |resource| resource.authorized?(:"#{:edit}_#{resource.class.name.pluralize.downcase}") } 
+    node(:canDelete) {|resource| resource.authorized?(:"#{:delete}_#{resource.class.name.pluralize.downcase}") } 
```
We should call this `can_destroy` so this matches the permissions name.

timogoebel commented on app/views/api/v2/authorizable.json.rabl:
```diff
@@ -0,0 +1,7 @@
+if @object.try(:class).try(:ancestors).to_a.include?(Authorizable)
+    node(:canEdit) { |resource| resource.authorized?(:"#{:edit}_#{resource.class.name.pluralize.downcase}") } 
+    node(:canDelete) {|resource| resource.authorized?(:"#{:delete}_#{resource.class.name.pluralize.downcase}") } 
+    node(:canCreate) { |resource| resource.authorized?(:"#{:create}_#{resource.class.name.pluralize.downcase}") } 
```
`can_create` might not make a lot of sense here. if the index action returns zero results you have no way to tell that the user is authorized to create more. :-)

ohadlevy replied:
agreed, it should probably passed as a prop once we have a full page react component (so you know if to add the create link or not)

timogoebel reviewed (dismissed):
@ripcurld: Thanks, some suggestions inline. Maybe we should move adding the permissions to the API to a separate PR for easier reviewing (and faster merging ;-)).

sharvit commented on app/views/models/index.html.erb:
```diff
@@ -2,32 +2,8 @@
 
-<table class="<%= table_css_classes 'table-fixed' %>">
-  <thead>
-    <tr>
```
IMO `request.query_parameters()` should be destruct to individual parameters so it will be easier to understand the flow and protect against unwanted parameters.

boaz0 replied:
done

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,63 @@
+
+class ModelsTable extends React.Component {
+  componentDidMount() {
+    this.props.updateSearch(this.props.data);
```
Could you destruct `data` and create a new object?
I feel like it's risky to pass the data object as is.

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,63 @@
+      sortBy,
+      sortOrder,
+    } = this.props;
+    return this.props.status === STATUS.PENDING ? (
```
Wouldn't it make more sense to pass a boolean `loading` prop and move the `this.props.status === STATUS.PENDING` logic to a redux selector and use it in `mapStateToProps`?

sharvit replied:
IMO I prefer it as well:

```js
if (loading) {
  return <LoadingTable />;
}

return <Table ... />;
```

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,63 @@
+    return this.props.status === STATUS.PENDING ? (
+      <div className="tableloader">
+        <Spinner size="lg" loading />
+      </div>
```
Could you move it into a new micro component?
Because the css is only relevant to this component, it can also move into this component and you may not need a ModelsTable.scss

amirfefer replied:
@waldenraines made a `LoadingState` component already in kattelo under `move_to_pf`

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,63 @@
+  }
+}
+
+ModelsTable.propTypes = {
```
You are missing some props.

boaz0 replied:
done.

sharvit replied:
You still missing the `status` prop.
Could you also add `defaultProps` for the not required props?

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,63 @@
+      </div>
+    ) : (
+      <Table
+        columns={schema({
```
IMO the schema function should be imported instead receive it as a prop.

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableActions.js:
```diff
@@ -0,0 +1,43 @@
+} from './ModelsTableConsts';
+import { ajaxRequestAction } from '../../redux/actions/common';
+
+export const updateSearch = (current) => {
```
The action name feels more like `initModelsTable`

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableActions.js:
```diff
@@ -0,0 +1,43 @@
+    results: current.table,
+  };
+
+  const payload = Object.keys(init)
```
IMO it feels hacky and it's difficult to follow and understand how the payload ends up.
Would it be easier if the action receive a destructed object?
```js
export const updateSearch = ({ page, perPage, order, searchQuery, table }) => { ... };
```

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableFormats.js:
```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+
+export const hostsCountFormat = (value, { rowData: { name } }) => (
```
I found out it's much easier to maintain if you separate your formaters into `component` and `formater` who only render the relevant component.
See:
https://github.com/Katello/katello/tree/master/webpack/move_to_foreman/components/common/table

boaz0 replied:
SGTM :+1:

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableReducer.js:
```diff
@@ -0,0 +1,47 @@
+  switch (action.type) {
+    case INIT_TABLE:
+      return Immutable.merge(state, {
+        ...action.payload,
```
IMO, better to destruct the payload and be in control about what's get into the store.

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableSchema.js:
```diff
@@ -0,0 +1,100 @@
+  ellipsisFormat,
+} from '../common/table';
+
+export const schema = ({
```
IMO it should be named `cretaeModelsTableSchema`.
Also the filename.

boaz0 replied:
TBH, I hate this function. I don't think it's readable and it makes me sad to think that I am building an abstraction over something that doesn't need it right now.

I am going to pull it back to the `<ModelsTable>` component. The reason for that is because the schema should be coupled to ModelsTable and the `<Table>` component gives a good abstraction over setting up a table view.

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/index.js:
```diff
@@ -0,0 +1,17 @@
+import reducer from './ModelsTableReducer';
+import * as actions from './ModelsTableActions';
+
+const mapStateToProps = ({ models }) => ({ ...models, schema });
```
IMO I prefer to pass individual props instead passing the whole state object. It help me to stay in control and I can understand the code better.

boaz0 replied:
You mean, to take only the relevant attributes from the state that the component needs, right?
If so, I agree, it does make it more readable.
Thanks a lot!

sharvit replied:
Yap :+1:

sharvit commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,48 @@
+
+const EmptyState = (props) => {
+  const {
+    icon = 'add-circle-o',
```
Could you add propTypes and default props?

sharvit replied:
And move the props destructing to `const EmptyState = ({ icon, ... }) => { ... };`

boaz0 replied:
Sure, no problem. I will check if `Katello` updated this file and add this if they don't have them too.

ripcurld replied:
Sorry but no test were found in Katello for this component... I will add them in the next days.

sharvit commented on webpack/assets/javascripts/react_app/components/common/emptyState/index.js:
```diff
@@ -0,0 +1,48 @@
+import React from 'react';
+import { EmptyState as PfEmptyState, Button } from 'patternfly-react';
+
+const EmptyState = (props) => {
```
Need testing?

boaz0 replied:
Yep, I missed the tests when I moved this component from `Katello`.
Thanks for noticing :100:

boaz0 replied:
Sorry no tests exist in `Katello` for this component, I will add some in the next days.

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/index.js:
```diff
@@ -0,0 +1,43 @@
+import { Table as PfTable } from 'patternfly-react';
```
Could you use the refactored table from katello?

boaz0 replied:
What do you mean by that :smile: ? This is the `<Table>` component that I took from `Katello`.

sharvit replied:
We refactor the table lately :disappointed: 

https://github.com/Katello/katello/pull/7467

sharvit reviewed (changes requested):
Awesome work @ripcurld :+1:

boaz0 commented:
I think this PR becomes bigger and bigger every iteration.
Unless all of you don't mind review over 1,000 LOC, I am not sure if it is a good idea to continue changing things here but to break it to small-ish changes and to focus here only on `Models Table`.
Probably, I will break it to:
 - API (already done in https://github.com/theforeman/foreman/pull/5814)
 - Table + Table Formatters
 - EmptyState

@ohadlevy @sharvit @waldenraines @timogoebel @amirfefer @tstrachota 

Thanks.

tstrachota commented:
Thanks for all your efforts @ripcurld! I definitely agree it would be better to split this PR into smaller parts. How about moving it to the "Labs" menu and develop it side by side with the existing hw models page? That would enable everyone to test it and we would replace the original page when it's completely ready.

ohadlevy commented:
@boaz1337 can you rebase (based on the assumption #5814 is going to be merged soon)? thanks

boaz0 commented:
@ohadlevy sure np...

mmoll commented:
ModelIntegrationTest failures related.

ohadlevy commented on app/views/models/index.html.erb:
```diff
@@ -2,32 +2,6 @@
-  </tbody>
-</table>
+<div id="models_table"></div>
+<%= mount_react_component("ModelsTable", "#models_table", {}.to_json) %>
```
is the empty hash required? I see the [default](https://github.com/theforeman/foreman/blob/develop/app/helpers/reactjs_helper.rb#L2) is an array - not sure if that's right or not..

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,152 @@
+      documentation: {},
+      action: {
+        url: '/models/new',
+        title: 'Create Hardware Model',
```
extract the string?

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,152 @@
+      order: this.props.sortOrder,
+    };
+
+    const schema = [
```
does it make sense to take out the schema outside of this render? its very hard to read when its as long.

boaz0 replied:
Yes, working on it.

boaz0 replied:
Done

sharvit replied:
Could you actually separate it into a different file?

It makes sense to me to put it in `ModelsTableSchema` with a main exported function call `createModelsTableSchema(getTableItems, sortBy, sortOrder)`.

You can also put there the other functions `sortableColumn`, `column`, `sortControllerFactory`.

I think it will make it easier to understand and follow.

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,152 @@
+    ];
+
+    if (this.props.search && !this.props.results.length) {
+      return <MessageBox key="models-table-no-results" icontype="info" msg="no match was found" />;
```
not sure if this brings a lot of value, as you already have that message as part of the pagination - e.g. 
![image](https://user-images.githubusercontent.com/20167/46917993-232dfc00-cfd5-11e8-9c36-b169abd4f264.png)

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,152 @@
+      return <MessageBox key="models-table-no-results" icontype="info" msg="no match was found" />;
+    }
+    return (
+      <Loader status={this.props.status}>
```
in my env, it was showing a spinner for a short time, i wonder if we should show a spinner only if we had like 200ms without a response or something along those lines...?

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,152 @@
+          <MessageBox
+            key="models-table-error"
+            icontype="error-circle-o"
+            msg="Oh no! Could not receive data"
```
extract?

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableActions.js:
```diff
@@ -0,0 +1,20 @@
+import { ajaxRequestAction } from '../../redux/actions/common';
+
+export const getTableItems = (controller, query) => (dispatch) => {
+  const url = new URI(`/api/${controller}`);
```
I'm OK with this file for now, but its semi reusable, e.g. the action names are specific (model) but the getTablesItem function is generic... would be nice to have reusable code? maybe we could have a TableRequestAction (similar to AjaxRquestAction)

boaz0 replied:
Yep, that's what I thought too. I will export it to common actions.

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/formatters/actionsCellFormatter.js:
```diff
@@ -0,0 +1,20 @@
+  (can_delete ? (
+    <span className="btn btn-sm btn-default">
+      <a
+        data-confirm={__(`Delete ${name}?`)}
```
at some point this needs to be converted into API DELETE, maybe add a TODO here?

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/formatters/actionsCellFormatter.js:
```diff
@@ -0,0 +1,20 @@
+        rel="nofollow"
+        data-method="delete"
+        href={`/${controllerPluralize}/${id}-${encodeURI(name)}`}>
+        Delete
```
not extracted

ohadlevy reviewed (changes requested):
looks good, some minor nit picks in code

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.test.js:
```diff
@@ -0,0 +1,15 @@
+import ModelsTable from './ModelsTable';
+
+describe('ModelsTable', () => {
+  it('smoke test', () => {
```
Isn't it necessary to add a snapshot test?

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableReducer.js:
```diff
@@ -0,0 +1,43 @@
+    case HW_MODEL_TABLE_FAILURE:
+      return state
+        .set('status', STATUS.ERROR)
+        .set('error', action.payload.error);
```
Don't you want to set the `results` back to empty-array when error appear?

boaz0 replied:
If there is an error, the results are not going to be shown.
In that case the user can only refresh the page which clears the Redux state.
Although, I don't see any point in doing so, maybe it makes the code more readable.

sharvit replied:
I see the state is our main source of truth and the place we go first when we want to understand/define what should actually get rendered.
I believe it must make sense so we can understand the application flow.

boaz0 replied:
Yep, I see your point here :dark_sunglasses: 
I updated the code.

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/formatters/actionsCellFormatter.js:
```diff
@@ -0,0 +1,22 @@
+export const deleteActionCellFormatter = controllerPluralize => (
+  _,
+  // eslint-disable-next-line camelcase
+  { rowData: { can_delete, name, id } },
```
You can get ride of the eslint comment by doing:
```jsx
{ rowData: { can_delete: canDelete, name, id } },
```

and use it as `canDelete`

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/formatters/nameCellFormatter.js:
```diff
@@ -0,0 +1,25 @@
+  {
+    rowData: {
+      // eslint-disable-next-line camelcase
+      can_edit, actions, id, name,
```
same here, can remove the comment by doing:
```jsx
can_edit: canDelete, actions, id, name,
```

sharvit reviewed (changes requested):
Thanks @boaz1337 :+1:

mmoll commented:
this needs a rebase and eslint fixes :)

boaz0 commented:
@ohadlevy @sharvit can you please take a look?

Thanks.

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,50 @@
+      getTableItems, sortBy, sortOrder, error,
+    } = this.props;
+
+    return this.props.status !== STATUS.ERROR ?
```
I assume you are avoiding the spinner / loader on purpose ?

boaz0 replied:
Yes. I would like to set a delay before the spinner will show. Otherwise I believe the user experience will be bumpy. Because it takes less than 100ms to show the data so the spinner is shown and quickly the table is rendered.

ohadlevy replied:
thanks to @xprazak2  we could use this once merged  https://github.com/patternfly/patternfly-react/pull/886/

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableReducer.js:
```diff
@@ -0,0 +1,31 @@
+    case 'MODELS_TABLE_REQUEST':
+      return state.set('status', STATUS.PENDING);
+    case 'MODELS_TABLE_SUCCESS':
+      return Immutable.merge(state, {
```
just wondering why not state.merge ?

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableReducer.js:
```diff
@@ -0,0 +1,31 @@
+      });
+    case 'MODELS_TABLE_FAILURE':
+      return state
+        .set('status', STATUS.ERROR)
```
can you please use merge?

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableSchema.js:
```diff
@@ -0,0 +1,65 @@
+  property: sortBy,
+  order: sortOrder,
+});
+
```
I have a feeling that everything above this line is generic and less relevant to the models page? I was just trying to understand when someone wants to reproduce another table, which part he actually is going to copy/paste?

boaz0 replied:
Yep that's probably a good idea to move it to `components/common/table` I think. Thanks for the review.

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/components/Table.js:
```diff
@@ -0,0 +1,51 @@
+  ...props
+}) => {
+  const body = children || [
+    <PfTable.Header key="header" />,
```
no empty state? (or is that currently handled by the server ATM?)

boaz0 replied:
Yes, currently it's handled by the server and that code isn't used.
I thought maybe when react-router will get in, it'll be the right time to add it.

WDYT?

ohadlevy replied:
I'm in favor, but would love to see this merged too, so maybe as a follow up PR? (or once we move all of the components into one HOC?)

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/formatters/actionsCellFormatter.js:
```diff
@@ -0,0 +1,21 @@
+) =>
+  // eslint-disable-next-line camelcase
+  (canDelete ? (
+    <span className="btn btn-sm btn-default">
```
use a Button component?

boaz0 replied:
I am not sure about this now. It appears that using `Button` instead of `span` changes the look of the action button.

ohadlevy reviewed (changes requested):
works well - thank you @boaz1337!

I've left a few minor issues / questions.
I also saw the following error in my console
```
The prop `children` is marked as required in `EllipisWithTooltip`, but its value is `null`.
    in EllipisWithTooltip (created by BodyRow)
    in BodyRow (created by Body)
    in tbody (created by Body)
    in Body (created by TableBody)
    in TableBody (created by Table)
    in table (created by Provider)
    in Provider (created by TablePfProvider)
    in TablePfProvider (created by Table)
    in div (created by Table)
    in Table (created by ModelsTable)
    in ModelsTable (created by Connect(ModelsTable))
    in Connect(ModelsTable)
    in Provider
```

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,50 @@
+
+class ModelsTable extends React.Component {
+  componentDidMount() {
+    this.props.getTableItems('models', new URI(window.location.href).query(true));
```
Could you move the `new URI(window.location.href).query(true)` into an helper file?
We might want to reuse it

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,50 @@
+
+  render() {
+    const {
+      getTableItems, sortBy, sortOrder, error,
```
I think it would be more readable if all the props would get extracted. (`status`, `results`)

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableSchema.js:
```diff
@@ -0,0 +1,38 @@
+  cellFormatter,
+} from '../common/table';
+
+const schema = (apiCall, by, order) => {
```
nitpick: Should we call it `createModelsTableSchema` since it's a function and not the actual schema?

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/formatters/hostsCountCellFormatter.js:
```diff
@@ -0,0 +1,12 @@
+  value,
+  { rowData: { name } },
+) => (
+  <a href={`hosts?search=${controllerSingular}+%3D+"${encodeURI(name)}"`}>
```
is this supposed to be generic? I'm concerned it won't always be possible to generate the search query in this way, and perhaps we should provide it as a parameter, ref to another function etc? I'm fine with doing this in a later PR.

boaz0 replied:
Yes, it's supposed to be generic.

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/formatters/actionsCellFormatter.js:
```diff
@@ -0,0 +1,21 @@
+  (canDelete ? (
+    <Button bsStyle="default">
+      <a
+        data-confirm={__(`Delete ${name}?`)}
```
Should we translate this string together with the `name`?
Or it's better to?
```jsx
data-confirm={`${__('Delete')} ${name}?`}
```

@ohadlevy?

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/formatters/actionsCellFormatter.js:
```diff
@@ -0,0 +1,21 @@
+  { rowData: { can_delete: canDelete, name, id } },
+) =>
+  (canDelete ? (
+    <Button bsStyle="default">
```
Should we separate the markup into a standalone component?
`deleteActionCellFormatter` that render `DeleteActionCell`.

So we also have proper prop-types and default-props.

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/schemaHelpers/index.js:
```diff
@@ -0,0 +1,54 @@
+  label,
+  mdWidth,
+  sortController,
+  additionalCellFmt = [],
```
nitpick: Could you call it `additionalCellFormatters`?

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/schemaHelpers/index.js:
```diff
@@ -0,0 +1,54 @@
+    property,
+    label,
+    [sortableHeaderFormatter(sortController), headerFormatterWithProps],
+    [...additionalCellFmt, ellipsisCellFormatter],
```
Do we want to use the `ellipsisCellFormatter` on each `sortableColumn` or it's the consumer choice?

boaz0 replied:
I have no idea. Maybe @ohadlevy can help with that.

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/schemaHelpers/index.js:
```diff
@@ -0,0 +1,54 @@
+  },
+});
+
+export const sortControllerFactory = (apiCall, sortBy, sortOrder) => ({
```
Could you separate each method to individual file?
I can see it grows.

sharvit reviewed (changes requested):
Great work @boaz1337 :+1: 
I like the way you kept everything small and reusable.

The Table is going to be used widely in our JS stack so it's is important to have a solid infrastructure we can scale up.

I want to suggest a separation between the `formatters` and the `components` they render:

```js
const HostsCountCell = ({ href, children }) => (
  <a href={href}>
    {children}
  </a>
);

const hostsCountCellFormatter = controllerSingular => (
  value,
  { rowData: { name } },
) => (
  <HostsCountCell href={`hosts?search=${controllerSingular}+%3D+"${encodeURI(name)}"`}>
    {value}
  </HostsCountCell>
);
```

* It will help us to achieve single responsibility for the `formatters`.
* This will also allow us to add `prop-types` and `default-props` to the components.
* It help us to write simpler tests.

Any toughts?

boaz0 commented:
Hi @sharvit 

I like your idea about separating components out of formatters and that's what I did.
Let me know if I missed something.

boaz0 commented:
Added more tests...

sharvit commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableSchema.js:
```diff
@@ -0,0 +1,38 @@
+  cellFormatter,
+} from '../common/table';
+
+const createModelsTableSchema = (apiCall, by, order) => {
```
Can you put JSDoc style comment to describe the parameters? It is not that clear to me.

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/formatters/actionsCellFormatter.js:
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import DeleteButton from '../components/deleteButton';
+
+export const deleteActionCellFormatter = controllerPluralize => (
```
Shoudn't you call this file `deleteActionCellFormatter.js`?

sharvit commented on webpack/assets/javascripts/react_app/redux/actions/common/getTableItemsAction.js:
```diff
@@ -0,0 +1,15 @@
+    dispatch,
+    requestAction: `${controller.toUpperCase()}_TABLE_REQUEST`,
+    successAction: `${controller.toUpperCase()}_TABLE_SUCCESS`,
+    failedAction: `${controller.toUpperCase()}_TABLE_FAILURE`,
```
I'm thinking about a way to reuse this logic (generating strings from a controller name) so reducers and tests could use it.

What do you think about seperating it to a different helpers?

```jsx
// helper
export const createTableActionTypes = controller => ({
  REQUEST: `${controller.toUpperCase()}_TABLE_REQUEST`,
  SUCCESS: `${controller.toUpperCase()}_TABLE_SUCCESS`,
  FAILURE: `${controller.toUpperCase()}_TABLE_FAILURE`,
});

// constants
export const MODELS_TABLE_CONTROLLER = 'models';

export const MODELS_TABLE_ACTION_TYPES = createTableActionTypes(MODELS_TABLE_CONTROLLER);
```

Then
* The component can use the `MODELS_TABLE_CONTROLLER` instead hardcoded.
* The reducer can use `MODELS_TABLE_ACTION_TYPES` instead hardcoded.
* It will be easier to scale up to different pages.

boaz0 replied:
I am not sure where I am using `MODELS_TABLE_ACTION_TYPES` though?

sharvit replied:
In the reducer. You are using strings atm

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/components/deleteButton.js:
```diff
@@ -0,0 +1,33 @@
+
+// TODO(bshuster): Move the confirmation to DialogModal that uses API to
+//                 delete the item.
+const DeleteButton = ({
```
Can you use CamelCase for the file name?

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/components/hostsCountCell.js:
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const HostsCountCell = ({
```
Can you use CamelCase for the file name?

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/components/nameCell.js:
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const NameCell = ({
```
Can you use CamelCase for the file name?

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/components/sortableHeader.js:
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const SortableHeader = ({ onClick, children, sortOrder }) => (
```
Can you use CamelCase for the file name?

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/schemaHelpers/column.js:
```diff
@@ -0,0 +1,19 @@
+export const column = (
```
Can you put JSDoc comment about `column` and it's parameters?

It's not that the code is not clear enaugh, but I think we can generate some cool docs later.

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/schemaHelpers/sortableColumn.js:
```diff
@@ -0,0 +1,35 @@
+} from '../formatters';
+import { column } from './column';
+
+export const sortableColumn = (
```
Can you put JSDoc comment?

sharvit commented on webpack/assets/javascripts/react_app/redux/actions/common/getTableItemsAction.js:
```diff
@@ -0,0 +1,15 @@
+import URI from 'urijs';
+import { ajaxRequestAction } from './index';
+
+export const getTableItems = (controller, query) => (dispatch) => {
```
Can you put JSDoc comment?

sharvit commented on webpack/assets/javascripts/react_app/components/common/table/schemaHelpers/sortableColumn.js:
```diff
@@ -0,0 +1,35 @@
+    { sort: true, sortDirection: '', className: `col-md-${mdWidth}` },
+  );
+
+export const sortControllerFactory = (apiCall, sortBy, sortOrder) => ({
```
Can you put JSDoc comment?

sharvit reviewed (changes requested):
Thanks @boaz1337, Amazing implementation :+1: 

I am very sorry i'm nitpiking, the reason is because this is going to be a key part in foreman and in plugins so we need to make sure everyone can understand, use it and scale it.
You done a great job with it and I think I could find a few other minur issues.

Thanks :1st_place_medal:

sharvit commented on webpack/assets/javascripts/react_app/redux/actions/common/getTableItemsAction.js:
```diff
@@ -0,0 +1,21 @@
+import URI from 'urijs';
```
I would prefer to put this file next to the table code.

Could you move it to `components/common/table/actions/.`?

sharvit replied:
bump?

sharvit reviewed (changes requested):
Thanks @boaz0 :+1: 

I got my last nitpick and it is ready to be merged from my side.

ohadlevy commented:
@boaz0 can you rebase please?

amirfefer reviewed (dismissed):
@boaz0  could you rebase this? I think it's ready to be merged :+1:

ohadlevy commented:
@timogoebel can you have another look if your requested changes are fulfilled ?  thanks :)

ohadlevy commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,57 @@
+
+class ModelsTable extends React.Component {
+  componentDidMount() {
+    this.props.getTableItems('models', getURIQuery(window.location.href));
```
I'm also missing a spinner on initial load?

ohadlevy replied:
I'm missing where you fixed this, can you give me a hint? thanks.

boaz0 replied:
No I saw you wanted it on initial load. I added a spinner every time API is being called.
I think I can fix this in one hour or less.

boaz0 replied:
Need a follow up: Pagination to be rendered inside the ModelsTable.
Right now it is rendered from `erb`. I am checking how much work it required. If it is not a lot, I will update this commit.

ohadlevy commented:
test failures not related https://community.theforeman.org/t/can-we-drop-node-js-6-14-from-travis-javascript-test-matrix/12737

boaz0 commented:
@sharvit @ohadlevy can you please take a look?
Thanks.

ohadlevy commented:
have a look at Gilad's audit commit, as he is pretty much doing something
similar.

On Thu, Feb 14, 2019 at 2:42 PM boaz0 <notifications@github.com> wrote:

> *@boaz0* commented on this pull request.
> ------------------------------
>
> In
> webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js
> <https://github.com/theforeman/foreman/pull/5755#discussion_r256816279>:
>
> > @@ -0,0 +1,57 @@
> +import React from 'react';
> +import PropTypes from 'prop-types';
> +import { Table } from '../common/table';
> +import { STATUS } from '../../constants';
> +import MessageBox from '../common/MessageBox';
> +import { translate as __ } from '../../common/I18n';
> +import createModelsTableSchema from './ModelsTableSchema';
> +import { getURIQuery } from '../../common/helpers';
> +
> +class ModelsTable extends React.Component {
> +  componentDidMount() {
> +    this.props.getTableItems('models', getURIQuery(window.location.href));
>
> Need a follow up: Pagination to be rendered inside the ModelsTable.
> Right now it is rendered from erb. I am checking how much work it
> required. If it is not a lot, I will update this commit.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5755#discussion_r256816279>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx9YpEzi18CWv4m6f4tWm4ChpltRfks5vNVnIgaJpZM4U8um9>
> .
>

ohadlevy commented:
@glekner can you review please? thanks.

timogoebel commented:
@ohadlevy: I trust you with this one. Feel free to merge whenever you think this is ready.

sharvit reviewed (dismissed):
LGTM

sharvit commented:
Test failures seem related

glekner commented on webpack/assets/javascripts/react_app/common/helpers.js:
```diff
@@ -72,6 +73,13 @@ export const translateObject = obj =>
+ * the parameters and the values are the parameters' values.
+ * @param {String} url - the URL
+ */
+export const getURIQuery = url => new URI(url).query(true);
```
We'll probably want to move all the URI helpers from `PaginationHelper` to `common/helpers` or another file in the future.

glekner commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,95 @@
+      );
+
+    return (
+      <Spinner size="lg" loading={this.state.initialLoading}>
```
I have mixed feelings about this.
I think its good to show a spinner for the initial loading, but it also shows when changing pages,
the page flickers and the table disappears. 

IMO, you should check the state, if the results array is empty, show a spinner, if not show the table. this way, there won't be any flickers, and I think you can remove the internal state of the component along with the `componentDidUpdate` and `stopSpinner`
```suggestion
    if (results.length > 0) return renderTable;
    return <Spinner size="lg" loading />;
```

glekner reviewed (changes requested):
Awesome work @boaz0 !
I believe we need to change `render` method of the table

ohadlevy commented:
can't we reuse https://github.com/patternfly/patternfly-react/pull/886/files
?

On Sun, Feb 17, 2019 at 12:09 PM Gilad Lekner <notifications@github.com>
wrote:

> *@glekner* requested changes on this pull request.
>
> Awesome work @boaz0 <https://github.com/boaz0> !
> I believe we need to change render method of the table
> ------------------------------
>
> In webpack/assets/javascripts/react_app/common/helpers.js
> <https://github.com/theforeman/foreman/pull/5755#discussion_r257496981>:
>
> > @@ -72,6 +73,13 @@ export const translateObject = obj =>
>
>   */
>
>  export const translateArray = arr => arr.map(str => __(str));
>
>
>
> +/**
>
> + * Return the query in URL as Objects where keys are
>
> + * the parameters and the values are the parameters' values.
>
> + * @param {String} url - the URL
>
> + */
>
> +export const getURIQuery = url => new URI(url).query(true);
>
>
> We'll probably want to move all the URI helpers from PaginationHelper to
> common/helpers or another file in the future.
> ------------------------------
>
> In
> webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js
> <https://github.com/theforeman/foreman/pull/5755#discussion_r257497444>:
>
> > +        />
>
> +      ) : (
>
> +        <React.Fragment>
>
> +          <Table
>
> +            key="models-table"
>
> +            columns={createModelsTableSchema(getTableItems, sortBy, sortOrder)}
>
> +            rows={results}
>
> +          />
>
> +          <div id="pagination">
>
> +            <Pagination data={pagination} />
>
> +          </div>
>
> +        </React.Fragment>
>
> +      );
>
> +
>
> +    return (
>
> +      <Spinner size="lg" loading={this.state.initialLoading}>
>
>
> I have mixed feelings about this.
> I think its good to show a spinner for the initial loading, but it also
> shows when changing pages,
> the page flickers and the table disappears.
>
> IMO, you should check the state, if the results array is empty, show a
> spinner, if not show the table. this way, there won't be any flickers, and
> I think you can remove the internal state of the component along with the
> componentDidUpdate and stopSpinner
> ⬇️ Suggested change
>
> -      <Spinner size="lg" loading={this.state.initialLoading}>
>
> +    if (results.length > 0) return renderTable;
>
> +    return <Spinner size="lg" loading />;
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5755#pullrequestreview-204553710>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOxw4qoDx93qx0bLSBNcqG7cq1Rd8-ks5vOSpzgaJpZM4U8um9>
> .
>

ohadlevy commented:
@boaz0 what did you decided to do with the PR? please let me know AFAIU the options are:
1. revert back to ERB generating 3 react components (search, table, pagination)
2. combine the table and pagination into one component?

thanks!

iNecas reviewed (dismissed):
Seems like Ohad's request for reusing the loading state was incorporated. Tested from user's perspective and works well.

I've hit some weird stuff when having Redux dev tools enabled (I guess it might be related to the fact we're mixing redux with turbolinks or something like that) but haven't observed this weirdness wiwhout the Redux dev tools.

Rebase needed (+ probably fixing some pending tests)

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/table/components/TableBody.js:
```diff
@@ -0,0 +1,29 @@
+  );
+};
+
+TableBody.propTypes = {
```
I'm not sure where this is coming from (pf?) but I get this error in my console:
```
checkPropTypes.js:19 Warning: Failed prop type: The prop `children` is marked as required in `EllipisWithTooltip`, but its value is `null`.
    in EllipisWithTooltip (created by BodyRow)
    in BodyRow (created by Body)
    in tbody (created by Body)
    in Body (created by TableBody)
    in TableBody (created by Table)
    in table (created by Provider)
    in Provider (created by TablePfProvider)
    in TablePfProvider (created by Table)
    in div (created by Table)
    in Table (created by ModelsTable)
    in LoadingState (created by ModelsTable)
    in ModelsTable (created by Connect(ModelsTable))
    in Connect(ModelsTable) (created by I18nProviderWrapper(Connect(ModelsTable)))
    in IntlProvider (created by I18nProviderWrapper(Connect(ModelsTable)))
    in I18nProviderWrapper(Connect(ModelsTable)) (created by StoreProvider(I18nProviderWrapper(Connect(ModelsTable))))
    in Provider (created by StoreProvider(I18nProviderWrapper(Connect(ModelsTable))))
    in StoreProvider(I18nProviderWrapper(Connect(ModelsTable))) (created by DataProvider(StoreProvider(I18nProviderWrapper(Connect(ModelsTable)))))
    in DataProvider(StoreProvider(I18nProviderWrapper(Connect(ModelsTable))))
```

boaz0 replied:
I couldn't reproduce it. I guess it's because I am not checking the `value` in `common/table/formatters/ellipsisCellFormatter.js`. Let me know if it is fixed.

ohadlevy commented:
@amirfefer @sharvit @Laviro  I've rebased this PR, can you please have another look prior to merge? I'm OK with small follow up PR's if you think that is required, but I think this PR is pretty much ready? thanks!

Ron-Lavi commented on app/views/models/index.html.erb:
```diff
@@ -2,32 +2,7 @@
-</table>
-<%= will_paginate_with_info @models %>
+<div id="models_table"></div>
+<%= mount_react_component("ModelsTable", "#models_table", {
```
could be added as a follow up PR,
Maybe create an helper file for it, e.g.: [search_bar_helper.rb](https://github.com/theforeman/foreman/blob/develop/app/helpers/search_bar_helper.rb)
and maybe use the [component wrapper](https://github.com/theforeman/foreman/pull/6142/files#diff-834531c2ec3feb85c8b9e09c36d93cbbR19) so props won't be accessed in react by the `data` object.

sharvit replied:
IMO we should use the component-wrapper only when we want to update those props with redux.
`data` prop became a standard for us to know those props came from server rendering and it feels useful to me.

Ron-Lavi replied:
@amirfefer what are you thoughts ?
can you share about the `flat` option in `mount_react_component`  as we discussed offline ?

amirfefer replied:
you can use `{flattenData: true}` in the componentRegistery,
I do agree with @sharvit on this, but if we use a component both server and client rendering (like charts) IMO it would be better to use the `flattenData`

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,84 @@
+
+class ModelsTable extends React.Component {
+  componentDidMount() {
+    this.props.getTableItems('models', getURIQuery(window.location.href));
```
we started to use more objects destruction in our components:
e.g: `const { getTableItems } = this.props`

if you want to get the query from the url, you could use the URI library which is already used in the project,
https://medialize.github.io/URI.js/docs.html#accessors-search

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -0,0 +1,84 @@
+      error,
+      status,
+      results,
+      data: { pagination },
```
as mentioned above, if we use the component wrapper, we won't need the data object here.

Ron-Lavi reviewed (commented):
Awesome work @boaz0 !!
Left some inline comments which I am fine with adding to a follow up PR :)

boaz0 commented:
@ohadlevy I will revert `models.html.erb` and @glekner will take it from there.

boaz0 commented:
@Laviro feel free to open tickets and work on those. I think this PR is old and bloated.

sharvit reviewed (approved):
LGTM! Thanks @boaz0 :1st_place_medal:

ohadlevy commented:
how is this different than other places we use the mounter? if its not
specific to this PR, lets do this in a follow up pr for all mount points?

On Thu, Feb 21, 2019 at 4:14 PM Avi Sharvit <notifications@github.com>
wrote:

> *@sharvit* commented on this pull request.
> ------------------------------
>
> In app/views/models/index.html.erb
> <https://github.com/theforeman/foreman/pull/5755#discussion_r258943831>:
>
> > -            hash_for_edit_model_path(:id => model).merge(:auth_object => model, :authorizer => authorizer) %></td>
> -        <td class="ellipsis"><%= model.vendor_class %></td>
> -        <td class="ellipsis"><%= model.hardware_model %></td>
> -        <td class='ra'><%= link_to hosts_count[model], hosts_path(:search => "model = \"#{model}\"") %></td>
> -        <td class="col-md-1">
> -          <%= action_buttons(display_delete_if_authorized(hash_for_model_path(:id => model).
> -            merge(:auth_object => model, :authorizer => authorizer),
> -            :confirm => _("Delete %s?") % model.name)) %>
> -        </td>
> -      </tr>
> -    <% end %>
> -  </tbody>
> -</table>
> -<%= will_paginate_with_info @models %>
> +<div id="models_table"></div>
> +<%= mount_react_component("ModelsTable", "#models_table", {
>
> IMO we should use the component-wrapper only when we want to update those
> props with redux.
> data prop became a standard for us to know those props came from server
> rendering and it feels useful to me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5755#discussion_r258943831>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx28izb1_tck4uOm2OdXZjFTKlGAnks5vPqnCgaJpZM4U8um9>
> .
>

ohadlevy commented:
test failures not related, thanks!

sharvit commented:
Thanks @boaz0 and everyone who contributed to this PR :heart:
